metrics/cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,226 @@
+package cloudwatch
+
+import (
+	"sync"
+
+	"time"
+
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/cloudwatch"
+	"github.com/aws/aws-sdk-go/service/cloudwatch/cloudwatchiface"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/metrics"
+	"github.com/go-kit/kit/metrics/generic"
+)
+
+// CloudWatch receives metrics observations and forwards them to CloudWatch.
+// Create a CloudWatch object, use it to create metrics, and pass those metrics as
+// dependencies to the components that will use them.
+//
+// To regularly report metrics to CloudWatch, use the WriteLoop helper method.
+type CloudWatch struct {
+	mtx        sync.RWMutex
+	namespace  string
+	svc        cloudwatchiface.CloudWatchAPI
+	counters   map[string]*Counter
+	gauges     map[string]*Gauge
+	histograms map[string]*Histogram
+	logger     log.Logger
+}
+
+// New returns a CloudWatch object that may be used to create metrics. Namespace is
+// applied to all created metrics and maps to the CloudWatch namespace.
+// Callers must ensure that regular calls to Send are performed, either manually or with one of the helper methods.
+func New(namespace string, logger log.Logger, svc cloudwatchiface.CloudWatchAPI) *CloudWatch {
+	return &CloudWatch{
+		namespace:  namespace,
+		svc:        svc,
+		counters:   map[string]*Counter{},
+		gauges:     map[string]*Gauge{},
+		histograms: map[string]*Histogram{},
+		logger:     logger,
+	}
+}
+
+// NewCounter returns a counter. Observations are aggregated and emitted once
+// per write invocation.
+func (cw *CloudWatch) NewCounter(name string) *Counter {
+	c := NewCounter(name)
+	cw.mtx.Lock()
+	cw.counters[name] = c
+	cw.mtx.Unlock()
+	return c
+}
+
+// NewGauge returns a gauge. Observations are aggregated and emitted once per
+// write invocation.
+func (cw *CloudWatch) NewGauge(name string) *Gauge {
+	g := NewGauge(name)
+	cw.mtx.Lock()
+	cw.gauges[name] = g
+	cw.mtx.Unlock()
+	return g
+}
+
+// NewHistogram returns a histogram. Observations are aggregated and emitted as
+// per-quantile gauges, once per write invocation. 50 is a good default value
+// for buckets.
+func (cw *CloudWatch) NewHistogram(name string, buckets int) *Histogram {
+	h := NewHistogram(name, buckets)
+	cw.mtx.Lock()
+	cw.histograms[name] = h
+	cw.mtx.Unlock()
+	return h
+}
+
+// WriteLoop is a helper method that invokes Send every
+// time the passed channel fires. This method blocks until the channel is
+// closed, so clients probably want to run it in its own goroutine. For typical
+// usage, create a time.Ticker and pass its C channel to this method.
+func (cw *CloudWatch) WriteLoop(c <-chan time.Time) {
+	for range c {
+		if err := cw.Send(); err != nil {
+			cw.logger.Log("during", "Send", "err", err)
+		}
+	}
+}
+
+// Send will fire an api request to CloudWatch with the latest stats for
+// all metrics.
```

Comment:
I thought it might be worth keeping this around as someone might have their own way they want to fire the api call.